### PR TITLE
[memory] Refactor memory module & add MMU64

### LINKED ISSUE #14: Implement memory management system
Need memory management system:
- Implement memory::MMU
- Refactor memory::PhysMemory

### PR DESCRIPTION
**Pull request by VasiliyMatr**
* Refactor memory::PhysMemory
* Add memory::MMU64
* Add memory::SimpleMemoryMapper
* Migrate code to new memory module API
* A few other fixes

### CODE REVIEW

File: memory/include/sim/memory/pte.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#ifndef INCL_MEMORY_PTE_HPP
+#define INCL_MEMORY_PTE_HPP
+
+#include "sim/common.hpp"
```
Comment (TheRedHotHabanero) [inline memory/include/sim/memory/pte.hpp]: USE COMMON STYLE FOR INCLUDE!!!!

File: memory/src/memory.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    case Mode::SV48:
+        return SV48_LEVELS;
+    case Mode::SV57:
+        return SV57_LEVELS;
+
+    default:
+        SIM_ASSERT(0);
+    }
+
+    SIM_ASSERT(0);
```
Comment (TheRedHotHabanero) [inline memory/src/memory.cpp:35]: redundant check

File: sim_app/src/sim_app.cpp [hunk 3]
```diff
@@ -1,3 +1,6 @@
+#include "sim/common.hpp"
+#include "sim/memory/common.hpp"
+#include "sim/simulator/sim_instr.hpp"
```
Comment (TheRedHotHabanero) [inline sim_app/src/sim_app.cpp]: same. please use <> instead of ""

File: sim_app/src/sim_app.cpp [hunk 4]
```diff
@@ -27,7 +30,9 @@ void dump_gpr_file(const gpr::GPRFile &gpr_file) {
 } // namespace
 
 int main() {
-    std::vector<InstrCode> binaryInstructions = {
+    const PhysAddr code_seg_base = 0x5000000000;
```
Comment (TheRedHotHabanero) [inline sim_app/src/sim_app.cpp]: maybe its better to use upper case for consts?

File: simulator/include/sim/simulator.hpp [hunk 5]
```diff
@@ -1,6 +1,8 @@
 #ifndef INCL_SIM_SIMULATOR_HPP
 #define INCL_SIM_SIMULATOR_HPP
 
+#include <vector>
```
Comment (TheRedHotHabanero) [inline simulator/include/sim/simulator.hpp]: vector is usls

File: memory/src/memory.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            }
+
+            // Write PTE
+            if (m_phys_memory.write(pte_pa, pte).status !=
+                PhysMemory::AccessStatus::OK) {
+                return MapStatus::PHYS_MEMORY_ERROR;
+            }
+
+            // Mapping created
+            if (i == 0) {
```
Comment (TheRedHotHabanero) [inline memory/src/memory.cpp:205]: Maybe put this check at the beginning of func?

File: memory/src/memory.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+                break;
+            }
+        }
+
+        // Leaf PTE is already set
+        if (i == 0) {
```
Comment (TheRedHotHabanero) [inline memory/src/memory.cpp:211]: It seems like we can return it directly in case above

**Review by TheRedHotHabanero — APPROVED** [review]
Ok, but please take a look at my comments

File: memory/src/memory.cpp (hunk 6 above)
Comment (VasiliyMatr) [reply memory/src/memory.cpp:205]: Page table walk with PTEs insertions is required prior this check

File: memory/src/memory.cpp (hunk 7 above)
Comment (VasiliyMatr) [reply memory/src/memory.cpp:211]: Above check will exit for loop and report success. I will make control flow clearer

File: memory/src/memory.cpp (hunk 2 above)
Comment (VasiliyMatr) [reply memory/src/memory.cpp:35]: It's just to be sure

**Pull request merged**